modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php

```diff
@@ -0,0 +1,355 @@
+			'href="%s" data-product_id="%s" data-product_sku="%s"',
+			esc_attr( $url ),
+			esc_attr( $product->get_id() ),
+			esc_attr( $product->get_sku() )
+		);
+		$url            = str_replace( 'href=', $new_attributes );
+		return $url;
+	}
+
+	/**
+	 * Track a product page view
+	 */
+	public function product_detail() {
+
+		global $product;
+		$blogid = Jetpack::get_option( 'id' );
+
+		$item_details = array(
+			'id'       => $product->get_id(),
+			'name'     => $product->get_title(),
+			'category' => Jetpack_WooCommerce_Analytics_Utils::get_product_categories_concatenated( $product ),
+			'price'    => $product->get_price(),
+		);
+		wc_enqueue_js(
+			"_wca.push( {
+				'_en': 'woocommerceanalytics_product_view',
+				'blog_id': '" . esc_js( $blogid ) . "',
+				'pi': '" . esc_js( $item_details['id'] ) . "',
+				'pn': '" . esc_js( $item_details['name'] ) . "',
+				'pc': '" . esc_js( $item_details['category'] ) . "',
+				'pp': '" . esc_js( $item_details['price'] ) . "',
+				'ui': '" . esc_js( $this->get_user_id() ) . "',
+			} );"
+		);
+	}
+
+	/**
+	 * On the Checkout page, trigger an event for each product in the cart
+	 */
+	public function checkout_process() {
+
+		$universal_commands = array();
+		$cart               = WC()->cart->get_cart();
+		$blogid             = Jetpack::get_option( 'id' );
+
+		foreach ( $cart as $cart_item_key => $cart_item ) {
+			/**
+			* This filter is already documented in woocommerce/templates/cart/cart.php
+			*/
+			$product = apply_filters( 'woocommerce_cart_item_product', $cart_item['data'], $cart_item, $cart_item_key );
+
+			$item_details = array(
+				'id'       => $product->get_id(),
+				'name'     => $product->get_title(),
+				'category' => Jetpack_WooCommerce_Analytics_Utils::get_product_categories_concatenated( $product ),
+				'price'    => $product->get_price(),
+				'quantity' => $cart_item['quantity'],
+			);
+
+			array_push(
+				$universal_commands,
+				"_wca.push( {
+					'_en': 'woocommerceanalytics_product_checkout',
+					'blog_id': '" . esc_js( $blogid ) . "',
+					'pi': '" . esc_js( $item_details['id'] ) . "',
+					'pn': '" . esc_js( $item_details['name'] ) . "',
+					'pc': '" . esc_js( $item_details['category'] ) . "',
+					'pq': '" . esc_js( $item_details['quantity'] ) . "',
+					'ui': '" . esc_js( $this->get_user_id() ) . "',
+				} );"
+			);
+		}
+
+		wc_enqueue_js( implode( "\r\n", $universal_commands ) );
+	}
+
+	/**
+	 * After the checkout process, fire an event for each item in the order
+	 *
+	 * @param string $order_id Order Id.
+	 */
+	public function order_process( $order_id ) {
+		$order              = wc_get_order( $order_id );
+		$universal_commands = array();
+		$blogid             = Jetpack::get_option( 'id' );
+
+		// loop through products in the order and queue a purchase event.
+		foreach ( $order->get_items() as $order_item_id => $order_item ) {
+			$product = $order->get_product_from_item( $order_item );
+
+			$item_details = array(
```

Comment:
Yeah looks like we could re-use this logic in a few places now, and `array_merge()` in other attributes like `order` here when needed.